### PR TITLE
[HandshakeToHW] Fix none-type usage in join and buffer

### PR DESCRIPTION
**Pull request by mortbopet**

### CODE REVIEW

File: lib/Conversion/HandshakeToHW/HandshakeToHW.cpp [hunk 1]
```diff
@@ -650,6 +650,7 @@ class HandshakeConversionPattern : public OpConversionPattern<T> {
     for (auto &input : inputs)
       valids.push_back(input.valid);
     Value allValid = s.bAnd(valids);
+    output.valid->setValue(allValid);
```
Comment (Dinistro) [inline lib/Conversion/HandshakeToHW/HandshakeToHW.cpp:653]: This change might require a removal of a similar line in `buildUnitRateLogic`: https://github.com/llvm/circt/blob/main/lib/Conversion/HandshakeToHW/HandshakeToHW.cpp#L727

File: lib/Conversion/HandshakeToHW/HandshakeToHW.cpp [hunk 2]
```diff
@@ -624,8 +624,8 @@ class HandshakeConversionPattern : public OpConversionPattern<T> {
       if (!isa<esi::ChannelType>(outputInfo.type))
         continue;
       OutputHandshake hs;
-      auto data = std::make_shared<Backedge>(
-          bb.get(cast<esi::ChannelType>(outputInfo.type).getInner()));
+      Type innerType = cast<esi::ChannelType>(outputInfo.type).getInner();
```
Comment (Dinistro) [inline lib/Conversion/HandshakeToHW/HandshakeToHW.cpp:627]: Nit: could we combine this with the`isa` above, e.g., with a `dyn_cast`.

File: lib/Conversion/HandshakeToHW/HandshakeToHW.cpp [hunk 3]
```diff
@@ -776,6 +777,8 @@ class JoinConversionPattern : public HandshakeConversionPattern<JoinOp> {
                    hw::HWModulePortAccessor &ports) const override {
     auto unwrappedIO = unwrapIO(s, bb, ports);
     buildJoinLogic(s, unwrappedIO.inputs, unwrappedIO.outputs[0]);
+    unwrappedIO.outputs[0].data->setValue(
```
Comment (Dinistro) [inline lib/Conversion/HandshakeToHW/HandshakeToHW.cpp:780]: As we do not write the value in `buildJoinLogic`, we shouldn't do it either for `buildForkLogic`. 
It's clear for a join that input's might have different types, and the same can hold or operations using fork logic.

**Review by Dinistro — COMMENTED** [review]
I added some small NITs, but this already looks very nice.

File: lib/Conversion/HandshakeToHW/HandshakeToHW.cpp (hunk 3 above)
Comment (Dinistro) [reply lib/Conversion/HandshakeToHW/HandshakeToHW.cpp:780]: Alternatively, we could add a `isControl` / `controlOnly` flag that would optionally drop the assignments of the data ports. "control" might be a bit misleading, though.

Comment (mortbopet) [reply lib/Conversion/HandshakeToHW/HandshakeToHW.cpp:780]: I'd be fine with leaving the data assignment out of `buildForkLogic` and let the caller handle how data is assigned. will change.

**Review by Dinistro — APPROVED** [review]
Could you add some tests for these operations as well? Having very small tests, like `HandshakeToFIRRTL` does, is very helpful to catch regressions. 

Edit: I guess you will do so in subsequent PRs.

**Comment by mortbopet** [comment]
> I guess you will do so in subsequent PRs.

Correct - plan is to slowly start migrating over HandshakeToFIRRTL tests once a minimally viable implementation has been reached.

**Comment by Dinistro** [comment]
We could already move the simple tests that only cover single operations, e.g., `test_join`, `test_fork`, etc.

**Pull request merged**